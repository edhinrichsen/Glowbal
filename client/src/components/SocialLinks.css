.socialLinks {
  border-left: var(--color-app) 1px solid;
  padding-left: 5px !important;
  margin-left: 5px;
  /* margin-top: -3px; */
  text-align: left;
}

.socialLinks a {
  color: var(--color-app);
}

.socialLinks ul {
  padding: 0;
  list-style-type: none;
}
.socialLinks li {
  box-shadow: var(--drop-shadow), 0px 0px 0px 1px #00000000;
  transition: color 0.3s, box-shadow 0.3s;
}

.socialLinks li:hover {
  color: var(--color-primary);
  box-shadow: var(--drop-shadow), 0px 0px 0px 1px var(--color-primary);
}

.exit-button-social-links {
  margin-top: -25px;
  margin-right: -50px;
  background: var(--del-gradient);
}

.skills-add {
  background-color: var(--background-color-primary) !important;
  box-shadow: var(--inner-shadow), 0px 0px 0px 3px var(--border-color), var(--drop-shadow) !important;
}

.ant-popover {
  width: 300px !important;
}
