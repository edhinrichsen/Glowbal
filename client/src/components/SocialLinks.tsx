import React, { useEffect, useState } from 'react';

import { Popover, Spin, Tooltip } from 'antd';
import {
  BehanceCircleFilled,
  TwitterCircleFilled,
  GithubFilled,
  DribbbleCircleFilled,
  LinkedinFilled,
  FacebookFilled,
  InstagramFilled,
  GlobalOutlined,
  EditOutlined,
} from '@ant-design/icons';

import './SocialLinks.css';
import TextInput from './TextInput';
import { useAuth0 } from '@auth0/auth0-react';

const API_URL = process.env.REACT_APP_API_URL
  ? process.env.REACT_APP_API_URL
  : 'https://localhost:5000/api/';

const SOCIAL_LINK_ICON: { [socialType: string]: JSX.Element } = {
  'github.com': <GithubFilled />,
  'twitter.com': <TwitterCircleFilled />,
  'behance.com': <BehanceCircleFilled />,
  'dribbble.com': <DribbbleCircleFilled />,
  'linkedin.com': <LinkedinFilled />,
  'facebook.com': <FacebookFilled />,
  'instagram.com': <InstagramFilled />,
};

type SocialLinksData = string[];

type SocialLinksProps = {
  /** userID to fetch social links from */
  userID?: string;
  /** Whether the links can be edited */
  isMyProfile: boolean;
};

function SocialLinks(props: SocialLinksProps) {
  const { getAccessTokenSilently, isAuthenticated } = useAuth0();
  const [isLoading, setIsLoading] = useState(true);

  const [links, setLinks] = useState<SocialLinksData>();

  async function fetchData(): Promise<void> {
    //setIsLoading(true);

    // If there is no userID, fetch own profile
    const route = props.isMyProfile ? 'links' : `links/${props.userID}`;

    // Call API
    try {
      const token = isAuthenticated ? await getAccessTokenSilently() : '';
      const res = await fetch(API_URL + route, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });

      // Check response is okay
      if (!res.ok) {
        console.error('Invalid response code', res.status, res.statusText);
        return;
      }

      // Check if response has profile data
      const resBody = await res.json();
      const data =
        'data' in resBody && 'links' in resBody['data']
          ? resBody['data']['links']
          : {};

      console.log('setting data...', { data });
      setLinks(data);

      setIsLoading(false);
    } catch (e) {
      if (setIsLoading) {
        setIsLoading(false);
      }
      const res = {
        status: 'error',
        message: [
          'Exception from fetch on client side (not API) - check if the API stopped running',
          e,
        ],
      };
      console.error(res, e);
    }
  }

  async function saveData(link: string): Promise<void> {
    //setIsLoading(true);

    // If there is no userID, fetch own profile
    const route = props.isMyProfile ? 'links' : `links/${props.userID}`;

    // Call API
    try {
      const token = await getAccessTokenSilently();
      const res = await fetch(API_URL + route, {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ data: { link: link } }),
      });

      // Check response is okay
      if (!res.ok) {
        console.error('Invalid response code', res.status, res.statusText);
        return;
      }

      fetchData();

      setIsLoading(false);
    } catch (e) {
      if (setIsLoading) {
        setIsLoading(false);
      }
      const res = {
        status: 'error',
        message: [
          'Exception from fetch on client side (not API) - check if the API stopped running',
          e,
        ],
      };
      console.error(res, e);
      //return res;
    }
  }

  async function deleteLink(link: string): Promise<void> {
    // If there is no userID, fetch own profile
    const route = props.isMyProfile ? 'links' : `links/${props.userID}`;

    // Call API
    try {
      const token = await getAccessTokenSilently();
      const res = await fetch(API_URL + route, {
        method: 'DELETE',
        headers: {
          Authorization: `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ data: { link: link } }),
      });

      // Check response is okay
      if (!res.ok) {
        console.error('Invalid response code', res.status, res.statusText);
        return;
      }

      fetchData();

      setIsLoading(false);
    } catch (e) {
      if (setIsLoading) {
        setIsLoading(false);
      }
      const res = {
        status: 'error',
        message: [
          'Exception from fetch on client side (not API) - check if the API stopped running',
          e,
        ],
      };
      console.error(res, e);
    }
  }

  const sanitisedLink = (link: string) => {
    return link
      .toLowerCase()
      .replace(/https?:\/\/(www\.)?/, '')
      .replace(/\/.*/, '');
  };

  useEffect(() => {
    fetchData();
    // eslint-disable-next-line
  }, []);

  if (isLoading) {
    return (
      <div className="socialLinks">
        <Spin />
      </div>
    );
  } else if (links?.length === 0 && !props.isMyProfile) {
    return <></>;
  } else {
    return (
      // <div className="socialLinks">
      <ul className="socialLinks">
        {links?.map((value, index) => {
          if (props.isMyProfile) {
            return (
              <Popover
                key={value}
                content={
                  <div>
                    <Tooltip title="Remove" placement="left">
                      <div
                        className="exit-button exit-button-social-links"
                        onClick={() => {
                          if (value === '') {
                            setLinks(links.filter((link) => link !== ''));
                          }
                          deleteLink(value);
                        }}
                      ></div>
                    </Tooltip>
                    <TextInput
                      placeholder="Enter social link"
                      editable={props.isMyProfile}
                      onChange={(newString: string) => {
                        saveData(newString);
                        deleteLink(value);
                      }}
                      text={value}
                    />
                  </div>
                }
                trigger="click"
              >
                {sanitisedLink(value) in SOCIAL_LINK_ICON ? (
                  <li>{SOCIAL_LINK_ICON[sanitisedLink(value)]}</li>
                ) : value === 'Enter URL' ? (
                  <li>
                    <EditOutlined />
                  </li>
                ) : (
                  <li>
                    <GlobalOutlined />
                  </li>
                )}
              </Popover>
            );
          } else {
            return (
              <Tooltip title={sanitisedLink(value)} placement="bottom">
                <a target="_blank" rel="noopener noreferrer" href={value}>
                  {sanitisedLink(value) in SOCIAL_LINK_ICON ? (
                    <li>{SOCIAL_LINK_ICON[sanitisedLink(value)]}</li>
                  ) : value === 'Enter URL' ? (
                    <></>
                  ) : (
                    <li>
                      <GlobalOutlined />
                    </li>
                  )}
                </a>
              </Tooltip>
            );
          }
        })}
        {props.isMyProfile ? (
          <Tooltip title={'Add Link'} placement="bottom">
            <li
              className="skills-add"
              onClick={() => {
                if (links) {
                  setLinks([...links!, '']);
                } else {
                  setLinks(['']);
                }
              }}
            >
              +
            </li>
          </Tooltip>
        ) : (
          <></>
        )}
      </ul>
    );
  }
}

export default SocialLinks;
