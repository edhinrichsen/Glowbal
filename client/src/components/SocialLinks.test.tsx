import React from 'react';
import { shallow } from 'enzyme';
import SocialLinks from './SocialLinks';

describe('Testing SocialLinks', () => {
  it('renders without crashing', () => {
    const component = shallow(<SocialLinks isMyProfile={false} userID={''} />);
    expect(component).toMatchSnapshot();
    component.unmount();
  });
});
