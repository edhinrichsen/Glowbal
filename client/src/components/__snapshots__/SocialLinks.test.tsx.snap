// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SocialLinks renders without crashing 1`] = `
<div
  className="socialLinks"
/>
`;
